Sensitivity component

rayjay00 opened the pull request:
Addresses issue #97 

This PR:
 * Adds sensitivity component and related story
 * Adds the component to `ProjectListItemCard`
 * **Added some styles/theme color configs** - This is where I'd like some feedback on this PR - primarily with how to add colors/reference them and any other best practices. See my comments in the code. Also, I'm not super familiar with TypeScript so extra eyes are appreciated. 👍 

**Screens**
![image](https://user-images.githubusercontent.com/4968115/80554653-d97cea80-899b-11ea-964b-12ae7c17d8ef.png)
![image](https://user-images.githubusercontent.com/4968115/80554664-e00b6200-899b-11ea-985b-3028d3839708.png)
![image](https://user-images.githubusercontent.com/4968115/80554675-e4d01600-899b-11ea-8fa3-0c380efc83fe.png)

rayjay00 commented on src/components/Sensitivity/Sensitivity.tsx:
```diff
@@ -0,0 +1,55 @@
+          return warning.main;
+        }
+
+        return '#8f928b';
```
This felt a little weird to add this color here. I need this light grey color for when the button says `Low` - is there a best practice to pull this in from the theme? It's used in a different spot, too, so I made a variable but couldn't quite pin down how to reference it. Any guidance is appreciated here!

CarsonF replied:
This is use case is a little tricky.
An argument could be made that these colors should just be defined here and not pulled from the theme. For example, we don't want to couple _High_ to our `secondary` color palette. That color could change and that doesn't mean we'd want this usage to change.
However, we also have `error` and `warning` definitions as well, which probably do mean the same thing as what we are trying to communicate here.

So it's more about mapping the intention of the theme option to the usage, rather than the color. If we can find a gray that makes sense in the theme and can be changed to the gray in designs and it still work across the app then let's do that. Otherwise we can just hardcode it here.

I'll dig more.

rayjay00 replied:
10-4

rayjay00 commented on src/theme/index.ts:
```diff
@@ -23,14 +25,18 @@ const createPalette = (options: PaletteOptions) => {
+      dark: midGrey,
+    },
+    warning: {
+      main: '#F2994A',
```
This is the orange color for when the label is `Medium`.

CarsonF replied:
I'm cool with this. But I want to double check with Patrick that he envisions this color for all warning like usages.

rayjay00 replied:
Ok sure! Happy to check in with Patrick on this.

rayjay00 commented on src/components/Sensitivity/Sensitivity.tsx:
```diff
@@ -0,0 +1,55 @@
+    <Box>
+      <Box className={iconWrapper}>
+        <VerifiedUserOutlined className={icon} color="disabled" />
+        <Typography variant="body2" color="textSecondary">
```
The color of this should be `#3C444E`. What's the best way to add that in to the theme config? `textSecondary` is a bit too light 🤷

CarsonF replied:
Based on the designs it looks like it's the primary text color.
That's defined here https://github.com/SeedCompany/cord-field/blob/d6e89115d4f8cf37cb75a69e47d273a9a1cbd0e8/src/theme/index.ts#L38
The color probably just changed in designs since I last looked.
I believe that's the default color for `Typography`, but don't take my word for it.

rayjay00 commented:
Seeing this in the CI - will resolve: 
![image](https://user-images.githubusercontent.com/4968115/80555145-670d0a00-899d-11ea-9e3f-a5a3888bfb64.png)

CarsonF commented on src/components/Sensitivity/Sensitivity.tsx:
```diff
@@ -0,0 +1,55 @@
+    chip: {
+      borderRadius: '4px',
+      color: 'white',
+      backgroundColor: ({ value }: { value: string }) => {
```
Ok so using property functions like this actually creates a new css class for each component that's mounted. Because of this I've come to the opinion that it's just easier to use inline `styles` instead of this. It's also easier on TypeScript, as you've seen a little bit of 😄

rayjay00 replied:
That is no bueno (but good to know)! I'm going inline.

CarsonF commented on src/components/Sensitivity/Sensitivity.tsx:
```diff
@@ -0,0 +1,55 @@
+import { VerifiedUserOutlined } from '@material-ui/icons';
+import { FC } from 'react';
+import * as React from 'react';
+import { Sensitivity as SensitivityType } from '../../api';
```
Hmm this is probably bad DX since we would be constantly fighting which import - the component or the type.
Since the type name is auto-generated, it would be better if we could come up with a different name for the component. Any ideas?

rayjay00 replied:
Agreed - thinking on it!

rayjay00 replied:
`SensitivityStatus`? 🤷

CarsonF replied:
Let's punt lol

CarsonF commented on src/components/Sensitivity/Sensitivity.tsx:
```diff
@@ -0,0 +1,55 @@
+  return (
+    <Box>
+      <Box className={iconWrapper}>
+        <VerifiedUserOutlined className={icon} color="disabled" />
```
Bringing up intent again, we aren't trying to communicate that this icon is disabled. Rather it's just a gray color.
It's possible that it matches our `textSecondary` color which would be a better fit.

CarsonF commented on src/components/Sensitivity/Sensitivity.tsx:
```diff
@@ -0,0 +1,55 @@
+  const { icon, iconWrapper, chip } = useStyles(props);
+
+  return (
+    <Box>
+      <Box className={iconWrapper}>
```
The `Box` component also defines a new css class for each component mounted, because of that I try to avoid them. Esp in cases where there could be lots of instance of the component on screen at once, like this.

So just replace with `<div>`.

CarsonF commented on src/components/Sensitivity/Sensitivity.tsx:
```diff
@@ -0,0 +1,55 @@
+      marginBottom: spacing(),
+    },
+    chip: {
+      borderRadius: '4px',
```
Let's keep an eye out. If other "chip" designs also have 4px radius then we should just move this to the theme.

CarsonF commented on src/theme/index.ts:
```diff
@@ -23,14 +25,18 @@ const createPalette = (options: PaletteOptions) => {
     },
     secondary: {
       main: '#ff5a5f',
+      dark: midGrey,
```
This doesn't make sense. light & dark should be different _shades_ of the same _hue_.

rayjay00 replied:
yeah this is an oxymoron! Thanks - will update to what it was before.

rayjay00 commented on src/components/Sensitivity/Sensitivity.tsx:
```diff
@@ -0,0 +1,55 @@
+  value: SensitivityType;
+}
+
+export const Sensitivity: FC<SensitivityProps> = (props) => {
+  const { icon, iconWrapper, chip } = useStyles(props);
```
I'll make this consistent with your way of doing it 👍

rayjay00 commented:
Ready for re-review. Just pushed a bunch of commits addressing all comments except those which will need Patrick's input and renaming the component (`SensitivityStatus` 🤷). Rather than styling inline, I went for adding a class conditionally since the colors were already in scope of the css. Let me know your thoughts. 👍 @CarsonF

CarsonF commented on src/components/Sensitivity/Sensitivity.tsx:
```diff
@@ -0,0 +1,62 @@
+
+  return (
+    <div>
+      <Box className={classes.iconWrapper}>
```
```suggestion
      <div className={classes.iconWrapper}>
```

CarsonF commented on src/components/Sensitivity/Sensitivity.tsx:
```diff
@@ -0,0 +1,62 @@
+export const Sensitivity: FC<SensitivityProps> = ({ value }) => {
+  const classes = useStyles();
+
+  const chipClass =
+    value === 'High'
+      ? classes.chipHigh
+      : value === 'Medium'
+      ? classes.chipMedium
+      : classes.chipLow;
```
There's a handy library that makes this kinda thing easier. 
```suggestion
  const chipClass = clsx({
    [classes.chip]: true,
    [classes.chipHigh]: value === 'High',
    [classes.chipMedium]: value === 'Medium',
    [classes.chipLow]: value === 'Low',
  });
```
Hmm same amount of lines. But another point I wanted to make is that the common chip class can be still be separated and multiple classes used.

rayjay00 replied:
Cool - I’ve this with the classNames package before - seems similar. Will add!

CarsonF commented on src/components/Sensitivity/Sensitivity.tsx:
```diff
@@ -43,19 +40,19 @@ export interface SensitivityProps {
-      : value === 'Medium'
-      ? classes.chipMedium
-      : classes.chipLow;
+  const chipClass = clsx({
+    [classes.chip]: true,
+    [classes.chipHigh]: value === 'High',
+    [classes.chipMedium]: value === 'Medium',
+    [classes.chipLow]: value === 'Low',
+  });
```
I thought of a better way to do this, if you want to tweak it again...
```suggestion
  const chipClass = clsx(classes.chip, classes[`chip${value}`]);
```

CarsonF commented on src/components/Sensitivity/Sensitivity.tsx:
```diff
@@ -0,0 +1,59 @@
+  return (
+    <div>
+      <div className={classes.iconWrapper}>
+        <VerifiedUserOutlined className={classes.icon} color="disabled" />
```
Change this to the secondary text color via css
